### ISSUE TITLE
Handle event queue: *in*correctness and deepcopy performance

### ISSUE THREAD

**Issue by vadcx** [Factorio-Biter-Battles/Factorio-Biter-Battles#492]
# History

Issue and reproduction test case: https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/issues/390

PR that introduced deepcopy to fix the _removal during iteration_ nil error: https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/pull/391

# Correctness

~~This is a short one~~, let me describe it first. _(yeah sure. short he said. it'll be fast he said)_

The code always does a full copy of the event handler (functions) array. Then it proceeds to iterate the original array, 1 through `#handlers`. Currently only `table.remove` is used to remove functions from the array, so it shouldn't cause any nil holes and misleading table lengths from the `#` operator. Good so far.

Only if the currently chosen handler happens to be nil, it checks the copy and tries to recover from the situation by running the copy's handler at that index.

https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/blob/master/utils/event_core.lua#L40-L49

```lua
local handlers_copy = table.deepcopy(handlers)
for i = 1, #handlers do
    local handler = handlers[i]
    if handler == nil and handlers_copy[i] ~= nil then
        if table.contains(handlers, handlers_copy[i]) then
            handler = handlers_copy[i]
        end
    end
    if handler ~= nil then
        xpcall(handler, errorHandler, event)
```

Situation 1: table.remove mid-iteration cannot cause nil in the middle at non-last index. Next.

Situation 2: we are at accessing index `#handlers` but in the middle of previous iteration it was removed with table.remove. I guess this is what had caused nil errors earlier. We recover successfully from the copy, asking what was at that index before. Correct, next.

Situation 3, to generalize the previous case: table.remove an index **ahead** of us or multiple indexes. We *attempt to* recover and call the now removed functions from copy, but end up repeating at least one index from copy, once too much.

Situation 4: table.remove on the current index of our iteration, shifting all next indices down. Result: it **skip** the event handler that was supposed to be next, but at the last index it calls the final index twice, once at `i-1` and once from copy at `copy[i]`.

Uhm now that I've just copy-pasted the call_handlers code and turned it into an example, all the code paths where something was removed seem to be wrong. Only tail additions without removals will work consistently by delaying until next event occurrence.

```lua
-- Copypasted from the table.lua library
----
function table.contains(t, e)
    return table.index_of(t, e) and true or false
end
function table.index_of(t, e)
    for k, v in pairs(t) do
        if v == e then
            return k
        end
    end
    return nil
end

----

function test_case_remove_previous_index(i)
    if i == 2 then
        table.remove(examples, i - 1)
    end
end

function test_case_remove_current_index(i)
    if i == 2 then
        table.remove(examples, i)
    end
end

function test_case_remove_next_index(i)
    if i == 2 then
        table.remove(examples, i + 1)
    end
end

function test_case_remove_last_index(i)
    if i == #examples-1 then
        table.remove(examples, #examples)
    end
end

examples = {"one", "two", "three", "four"}
examples_copy = {"one", "two", "three", "four"}

for i = 1, #examples do
    local sample = examples[i]
    if sample == nil and examples_copy[i] ~= nil then
        if table.contains(examples, examples_copy[i]) then
            sample = examples_copy[i]
        end
    end
    if sample ~= nil then
        -- executes handler in current code, thus manipulation must happen below
        print(sample)
        
        -- we can manipulate "examples" table here.
        test_case_remove_previous_index(i) -- one two four four
        --test_case_remove_current_index(i) -- one two four four
        --test_case_remove_next_index(i) -- one two four four
        --test_case_remove_last_index(i) -- nil handler index 4
    else
        print('nil handler', "index: ", i)
    end
end
```

This is worse than I thought, there's no way it handles anything correctly right now. OTOH additions/removals are very rare in the codebase.

# Performance cost of deepcopy

The way the profiler log is sorted made me notice the time taken by deepcopy only at the very end. I reran the profiler today.

**TLDR:** The deepcopy costs 5107ms/8750ms = 58.3% (opt_deepcopy) or 5126ms/8761ms = 58.5% (vanilla_deepcopy) **of total script running time** over a minute. Or 8.5% of a full minute aka roughly 5 UPS once we hit 16.6ms Update Time.

**What to do about it?** The `table.deepcopy` must be removed, event addition/removal must be non-invasive to the handler table while it's being iterated. Consider if event removal must happen immediately or can be scheduled until next occurrence. Thus maybe two functions: schedule register / now register & removal.

If I had the slightest idea of the structure of the event handlers and the library that does registrations/removal, I'd have yoinked that crap (deepcopy) right there and then. I have started at `event.lua` for an hour and couldn't *really* understand what ends up where. Why is there the Token system that claims local upvalues are desyncy, but is exactly how it operates. Why is there hot code compilation in  `add_handlers` and `add_removable_function`.

I 100% agree with [BigZet to eliminate handler element shifting](https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/pull/391#issuecomment-1763191888). But I don't comprehend the other code around events.

---

_One minute long profiler output when I had 15 FPS and barely 60 UPS from the game at 2024-04-14 17:05 UTC._ but running Factorio 1.1.106 with it.

**Profiler: Total Time vs Self-Time:** total time means the time of the current function itself PLUS all the children it has called. Self-Time is supposed to only represent the time the current function itself took. I find Total Time easier to navigate. Keep in mind that the profiler log doesn't stick to these names.

`/startProfiler` -> total time in log, whereas `/startProfiler anyWord` -> self-time in log

```
62313x "call_handlers" in ".../utils/event_core.lua", line 79. Total Elapsed: 8290.423145ms
-> calls 62313x "vanilla_deepcopy" in ".../utils/event_core.lua", line 31. Total Duration: 5126.081241ms
```

I am attaching all four logs here. `profiler-vanilla-deepcopy-totaltime.log` is preferred.

[profiler-vanilla-deepcopy-totaltime.log](https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/files/15087438/profiler-vanilla-deepcopy-totaltime.log)
[profiler-vanilla-deepcopy-selftime.log](https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/files/15087436/profiler-vanilla-deepcopy-selftime.log)
[profiler-optimized-deepcopy-selftime.log](https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/files/15087433/profiler-optimized-deepcopy-selftime.log)
[profiler-optimized-deepcopy-totaltime.log](https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/files/15087434/profiler-optimized-deepcopy-totaltime.log)

`vanilla_deepcopy` is [this for reference](https://github.com/wube/factorio-data/blob/5265f5c86fb0114cc4e19e88dfb8eba3480b68d1/core/lualib/util.lua#L6).

`opt_deepcopy`: I thought internal creation of temp function was the expensive part but it **surprised** me, it made no difference here.

```lua
local opt_deepcopy
do
    local function _copy(object, lookup_table)
        if type(object) ~= "table" then
            return object
            -- don't copy factorio rich objects
        elseif object.__self then
            return object
        elseif lookup_table[object] then
            return lookup_table[object]
        end
        local new_table = {}
        lookup_table[object] = new_table
        for index, value in pairs(object) do
            new_table[_copy(index, lookup_table)] = _copy(value, lookup_table)
        end
        return setmetatable(new_table, getmetatable(object))
    end

    opt_deepcopy = function(object)
        local lookup_table = {}
        return _copy(object, lookup_table)
    end
end
```

PS: Fixing the deepcopy will probably solve most of the freeze issues on map restart when people get dropped due to slow computers, because all the terrain copy stuff goes through event handlers.

**Comment by clifffrey** [comment]
This is an awesome issue description!

Just to brainstorm a possible solution, would something like the following work?

1. Try and optimize the event-handling code path, at the cost of making registering/unregistering event handlers more expensive
2. To do this, add variables along the lines of `bool currently_running_event_handlers` and `optional<EventHandler[]> new_eventhandlers`, and if add/remove run while `currently_running_event_handlers` is set to true, then do the deep copy into `new_eventhandlers` (unless it is already set) and then manipulate `new_eventhandlers`
3. After running event handlers, set `currently_running_event_handlers = false`, and if `new_eventhandlers` is set, replace the eventhandlers with it.

Are you planning on working on a solution to this?

**Comment by vadcx** [comment]
I suppose that's one way to do it. Flags wouldn't be my go to here (https://news.ycombinator.com/item?id=18442637) but whatever works. I had been thinking of something like TheBigZet suggested, a sort of in-place deletion and if we are to allow same-tick registration vs scheduled registration then possibly a second table to not mix up stuff.

If placing a "nil" somewhere in array part breaks it up (judging by side-effects it does turn 1-#tbl into array and the cut off rest is moved to hashtable?) then the values must be in-place replaced with dummies and removed in a second pass. But in Factorio they say the first 1024 elements are always ordered for pairs, not sure what this implies.

I thought someone with more experience on this part of event code might look at it too, otherwise I will try it myself again.

**Comment by clifffrey** [comment]
Just to clarify my suggestion slightly, it would involve something like adding global variables to event_core.lua of

```
-- map of event_name to handlers[]
local event_handlers = {}
-- set to an event_name if one is currently running
local currently_running_handler = nil
-- if non-nil, the replacement value for event_handlers[current_running_handler] once event handlers finish
local new_handlers_for_currently_running_handler = nil
```

I feel like any other solution will look like black-magic in how it manipulates `event_handlers[x]` while we are also iterating over it at the same time... but maybe I just don't see the beauty of the possible solution.

**Comment by morganchristiansson** [comment]
I also found deepcopy in my profiles.

Here is extract:
```
5056x "anonymous" in "__level__/utils/event_core.lua", line 56. Total Elapsed: 2231.494371ms
	5056x "call_handlers" in "__level__/utils/event_core.lua", line 36. Total Elapsed: 2150.956916ms
		13137x C function "xpcall". Total Elapsed: 985.638357ms
			<snip>
		5056x "deepcopy" in "__core__/lualib/util.lua", line 6. Total Duration: 1022.622504ms
```

We spent more time in deepcopy (1022ms) than in actual event handlers (985ms).
1 second is 1.6% of 1 minute. (1/60=0.016).

**Comment by morganchristiansson** [comment]
Can close this now that #681 is merged.
